docs/spec/auth/subkeys_proposal.md

```diff
@@ -0,0 +1,156 @@
+
+
+## Construction
+
+To begin, we define the struct that stores the metadata about a certain account.  This includes the pubkey itself, the routes it's permitted to sign transactions for, its daily allowance, and how much of its allowance has been used in the last 24 hour window.
+
+```golang
+type SubKeyMetadata struct {
+  PubKey               sdk.AccPubKey
+  PermissionedRoutes   []string
+  DailyFeeAllowance    sdk.Coins
+  DailyFeeUsed         sdk.Coins
+}
+```
+
+We create a new `Account` type called `SubKeyAccount` that stores a list of `SubKeyMetadata`s.
+
+```golang
+type SubKeyAccount struct {
+  Address       AccAddress
+  Coins         Coins
+  PubKey        sdk.AccPubKey
+  AccountNumber uint64
+  Sequence      uint64
+  SubKeys       []SubKeyMetadata
+}
+```
+
+In the `StdSignature` type we need to add a new field called `PubKeyIndex` which tells the anteHandler which SubKey to use to verify this signature.
+
+```golang
+// StdSignature represents a sig
+type StdSignature struct {
+	crypto.PubKey                  `json:"pub_key"` // optional
+  Signature     []byte           `json:"signature"`
+  PubKeyIndex   uint
+}
+```
+
+If a `StdSignature.PubKeyIndex` is 0, this means that the account's "master" PubKey should be used to verify the signature.  If the `StdSignature.PubKeyIndex` is `> 0`, it verifies the signature using the corresponding **1-indexed** SubKey in the SubKeyAccount.SubKeys slice.  Note, the SubKeys slice is 1-indexed rather than 0-indexed because the 0 index is used to refer to the master PubKey.  See pseudocode below:
+
+```
+if stdSig.PubKeyIndex == 0 {
+  verify stdSig using acc.PubKey
+} else {
+  verify stdSig using acc.SubKeys[stdSig.PubKeyIndex - 1]
+}
+```
+
+In the AnteHandler, if a transaction comes from a SubKey, it verifies the msg route is permitted.
+
+```
+if msg.Route not in subKeyMetadata.PermissionedRoutes {
+  return ErrNotPermitted
+}
+```
+
+When a transaction signed by a SubKey is accepted and has paid fees, we need to log the fees it has paid.  We increase the `SubKeyMetadata.DailyFeeUsed` field.
+
+If a new transaction from a subkey comes in and the `tx.Fee + subKeyMetadata.DailyFeeUsed > subKeyMetadata.DailyFeeAllowance`, then the transaction is rejected as the subkey is trying to exceed its daily fee allowance.
+
+But now we need a way to decrease the `DailyFeeUsed` field once transactions are past the 24 hour window.  To do this we us a time based iterator similar to the ones used in the governance and staking queues.
+
+We start by defining a new struct called a `DailyFeeSpend`, which is what will be inserted into the Queue.
+
+```golang
+type DailyFeeSpend struct {
+  Address              sdk.AccAddress
+  SubKeyIndex          uint
+  FeeSpent             sdk.Coins
+}
+```
+
+When a transaction using a SubKey has paid its fee, we insert a new entry into this queue's store.
```

Comment:
What kind overhead will this be adding? Both space and time.

Reply:
It would add a constant time operation for each transaction that uses a SubKey.

It will also add a constant size state insertion, but this only lasts 24 hrs, and then is pruned from state.